### ISSUE TITLE
Is there a way to set the random seed in MMDetection 3.0?

### ISSUE THREAD

**Issue by AndreaPi** [open-mmlab/mmdetection#10559]
When I train the same model twice (same config file, same dataset, same Docker container) the results are not the same, probably because the `numpy` random seed changes, according to the training log:

**Run 1**
```
------------------------------------------------------------
System environment:
    sys.platform: linux
    Python: 3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0]
    CUDA available: True
    numpy_random_seed: 797778965
    GPU 0: NVIDIA A100-SXM4-40GB
    CUDA_HOME: /usr/local/cuda
    NVCC: Cuda compilation tools, release 11.1, V11.1.105
    GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
    PyTorch: 1.9.0
.
.
```

**Run 2**
```
2023/06/26 09:44:47 - mmengine - INFO - 
------------------------------------------------------------
System environment:
    sys.platform: linux
    Python: 3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0]
    CUDA available: True
    numpy_random_seed: 1892126386
    GPU 0: NVIDIA A100-SXM4-40GB
    CUDA_HOME: /usr/local/cuda
    NVCC: Cuda compilation tools, release 11.1, V11.1.105
    GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
    PyTorch: 1.9.0
.
.
```

1. Is there a way to set the random seed from the `config` file? I had a look at the MMDetection 3.0 documentation, but I didn't find any entry for setting random seeds.
2. If there's no way from the `config` file, is it possible to set it by editing `tools/train.py `? What should I edit, then?

**Comment by hhaAndroid** [comment]
If you want to set the random seed during training, you can use the following command.

```shell
python ./tools/train.py \
    ${CONFIG} \                               # path of the config file
    --cfg-options randomness.seed=2023 \      # set seed to 2023
    [randomness.diff_rank_seed=True] \        # set different seeds according to global rank
    [randomness.deterministic=True]           # set the deterministic option for CUDNN backend
# [] stands for optional parameters, when actually entering the command line, you do not need to enter []
```

`randomness` has three parameters that can be set, with the following meanings.

- `randomness.seed=2023`, set the random seed to 2023.
- `randomness.diff_rank_seed=True`, set different seeds according to global rank. Defaults to False.
- `randomness.deterministic=True`, set the deterministic option for cuDNN backend, i.e., set `torch.backends.cudnn.deterministic` to True and `torch.backends.cudnn.benchmark` to False. Defaults to False. See https://pytorch.org/docs/stable/notes/randomness.html for more details.

**Comment by AndreaPi** [comment]
@hhaAndroid very nice! Rather than setting the options in the` train.py` call, I would like to set them in the config file for reproducibility.  Will it work if I add the following lines to my config.py?

```
randomness = dict(
    seed = 2023,
    diff_rank_seed=True,
    deterministic=True
)
```

**Comment by emvollmer** [comment]
Hi everyone, I'm just going to join this discussion with a question of my own! 
@hhaAndroid with regards to what you wrote. Setting the `randomness.seed` works fine for me, but when I try to set `randomness.deterministic=True` I always get the same error during training:
```
File "/path/to/venv/lib64/python3.8/site-packages/mmdet/models/dense_heads/anchor_head.py",
line 284, in _get_targets_single
bbox_weights[pos_inds, 2:  :] = 1.0

RuntimeError: linearIndex.numel()*sliceSize*nElemBefore ==
expandedValue.numel() INTERNAL ASSERT FAILED at
"../aten/src/ATen/native/cuda/Indexing.cu":384, please report a bug to PyTorch.
number of flattened indices did not match number of elements in the value
tensor: 132 vs 33
```
Do you know where this might be coming from? I've tried multiple different variations (with / without `seed` and / or `diff_rank_seed` definition) to no effect.

And @AndreaPi yes you can, check out the [MMEngine docs](https://github.com/open-mmlab/mmengine/blob/41c60cc3aab2064a8c5bff29acd265194fb5a705/docs/en/migration/runner.md#set-random-seed). I added my randomness definition into the `default_runtime.py` - I think I saw somewhere that's the place to put it but you could add it in somewhere else as well.

**Comment by AndreaPi** [comment]
@emvollmer thabks for the answer, but once I verify that your suggestion works, I'll close the issue, so maybe it would be better for you to open a new issue

**Comment by emvollmer** [comment]
@AndreaPi of course, thanks for the heads up. I assume you've already tried setting the options directly via `cfg-options` flag in the `train.py` call and haven't encountered the error I described?

**Comment by lllsgq** [comment]
> @AndreaPi当然，感谢您的提醒。`cfg-options`我假设您已经尝试直接通过调用中的标志设置选项`train.py`，并且没有遇到我描述的错误？
Hello, I've encountered the same issue as you. Did you manage to resolve it

**Comment by emvollmer** [comment]
@lllsgq If you're referring to my problem: No, sadly I haven't managed to solve it. But to be honest I have stopped working with MMDetection at the moment because of a different issue I can't find a solution for... In case you're successful, I'd be interested to hear it.

**Comment by lllsgq** [comment]
> @lllsgq如果您指的是我的问题：不，遗憾的是我还没有设法解决它。但说实话，我目前已经停止使用 MMDetection，因为我找不到解决方案的不同问题......如果你成功了，我很想听听它。
Thank you for your response. I found that it only works when I keep the seed set as 'seed = 2023,' but it seems that the experimental results can't be fully replicated.
